File: version1/webstats.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>

/* For char array initialization */
#define MAX_LINE_SIZE 4096

/* Used for simple double array cache */
static const int LBYTES = 2;
static const int TBYTES = 4;
static const int LGETS = 8;
static const int TGETS = 16;
static const int FGETS = 32;
static const int LFGETS = 64;

/* Used to manage fields on parsing */
static const int ADDRESS_FIELD = 0;
static const int DATE_FIELD = 3;
static const int HTTP_STATUS_CODE_FIELD = 8;
static const int BYTES_DOWNLOADED_FIELD = 9;

/* Used as delimeter for line parsing */
static const char *DELIMETER = " []\"";

/* Struct to hold webstats data */
struct stats {
	double local_bytes;
	double total_bytes;
	double local_gets;
	double total_gets;
	double failed_gets;
	double local_failed_gets;
	pthread_mutex_t mutex;
} webstats;

/*
 * parse_line(): Parse one line into string tokens separated by the given delimiters.
 * Beware: The input argument line is consumed in the process.
 *
 *	input:
 * 		line: A character array containing the current server log entry line
 * 		delim: A character array containing the delimiters to be used to separate fields in the line
 *      cache: A double array initalized to at least LFGETS+1 to cache critical data
 *      date: An empty character array to write a date to
 *
 * 	output:
 * 		cache: An double array with updated webstats data
 *      date: A character array with the extracted date written to it
 */
static void parse_line(char *line, const char *delim, double *cache, char *date)
{
	/* Setup Data */
	char *next;
	char *save;
	char *address;
	char *httpStatusCode;
	int cnt = 0;
	int bytes_downloaded = 0;

	/* Increment the total gets */
	cache[TGETS]++;

	/* Begin tokenization of the line */
	next = strtok_r(line, delim, &save);
	while(next) {
		/* Get our address */
		if(cnt == ADDRESS_FIELD) {
			address = (char *) malloc(strlen(next) + 1);
			strcpy(address, next);
		}

		/* Get and write the date */
		else if(cnt == DATE_FIELD)
			strcpy(date, next);

		/* Get our status code */
		else if(cnt == HTTP_STATUS_CODE_FIELD) {
			httpStatusCode = (char *) malloc(strlen(next) + 1);
			strcpy(httpStatusCode, next);

			/* If failure, we need to update our failed gets */
			if(atoi(httpStatusCode) == 404)
				cache[FGETS]++;
		}

		/* Get the bytes downloaded */
		else if(cnt == BYTES_DOWNLOADED_FIELD) {
			bytes_downloaded = atoi(next);

			/* Update the total bytes */
			cache[TBYTES] += bytes_downloaded;

			/* See if it's ours */
			if((strstr(address, "boisestate.edu") != NULL) || (strstr(address, "132.178") != NULL)) {
				/* Update our local gets and bytes downloaded */
				cache[LGETS]++;
				cache[LBYTES] += bytes_downloaded;

				/* Update our local failures if needed */
				if(atoi(httpStatusCode) == 404)
					cache[LFGETS]++;
			}

			/* Keep it clean */
			free(address);
			free(httpStatusCode);
			return;
		}
		next = strtok_r(NULL, delim, &save);
		cnt++;
	}

	/* Keep it clean in case we break out or miss the return above */
	free(address);
	free(httpStatusCode);
}

/**
 * initialize_webstats(): Initialize the webstats structure.
 */
static void initialize_webstats()
{
	webstats.total_bytes = 0;
	webstats.total_gets = 0;
	webstats.local_bytes = 0;
	webstats.local_gets = 0;
	webstats.failed_gets = 0;
	webstats.local_failed_gets = 0;
	pthread_mutex_init(&(webstats.mutex), NULL);
}


/**
 *	update_webstats(): Updates the webstats structure based on the input fields of current line.
 *
 *	input:
 *		num: The number of fields in the current webserver log entry
 *		field: An array of num strings representing the log entry
 */
static void update_webstats(double *cache)
{
	/* Lock out other threads */
	pthread_mutex_lock(&(webstats.mutex));

	/* Update our critical data */
	webstats.local_bytes += cache[LBYTES];
	webstats.total_bytes += cache[TBYTES];
	webstats.local_gets += cache[LGETS];
	webstats.total_gets += cache[TGETS];
	webstats.failed_gets += cache[FGETS];
	webstats.local_failed_gets += cache[LFGETS];

	/* Unlock so other threads get a turn */
	pthread_mutex_unlock(&(webstats.mutex));
}


/**
 * 	print_webstats(): Print out webstats on standard output.
 */
static void print_webstats()
{
	printf("%10s %15s   %15s  %15s\n", "TYPE", "gets", "failed gets", "MB transferred");
	printf("%10s  %15.0f  %15.0f  %15.0f\n", "local", webstats.local_gets,
	       webstats.local_failed_gets, (double) webstats.local_bytes / (1024 * 1024));
	printf("%10s  %15.0f  %15.0f  %15.0f\n", "total", webstats.total_gets,
	       webstats.failed_gets, (double) webstats.total_bytes / (1024 * 1024));
}


/**
 * process_file(): The main function that processes one log file
 *
 *	input:
 * 		ptr: Pointer to log file name.
 */
void *process_file(void *ptr)
{
	/* Setup our data */
	char *filename = (char *) ptr;
	char *linebuffer = (char *) malloc(sizeof(char) * MAX_LINE_SIZE);
	char *date = (char *) malloc(sizeof(char) * MAX_LINE_SIZE);
	double *cache = (double *) calloc(LFGETS + 1, sizeof(double));
	FILE *fin = fopen(filename, "r");

	/* Handle error of opening file */
	if(fin == NULL) {
		fprintf(stderr, "Cannot open file %s\n", filename);
		exit(1);
	}

	/* Get us that file data */
	char *s = fgets(linebuffer, MAX_LINE_SIZE, fin);

	/* Process the line if file had data */
	if(s != NULL) {
		/* Strip the first line for start date */
		parse_line(linebuffer, DELIMETER, cache, date);
		printf("Starting date: %s\n", date);

		/* Nab the rest of it all */
		while(fgets(linebuffer, MAX_LINE_SIZE, fin) != NULL) {
			parse_line(linebuffer, DELIMETER, cache, date);
			strcpy(linebuffer, "");
		}
		printf("Ending date: %s\n", date);

		/* Write to webstats (only ONCE to reduce critical code access) */
		update_webstats(cache);
	}

	/* Keep it clean */
	fclose(fin);
	free(date);
	free(linebuffer);
	free(cache);

	return NULL;
}

/**
 * Main insertion
 */
int main(int argc, char **argv)
{
	/* Check args */
	if(argc <  2) {
		fprintf(stderr, "Usage: %s <access_log_file> {<access_log_file>} \n", argv[0]);
		exit(1);
	}

	/* Setup data */
	int i;
	int numThreads = argc - 1;
	char *program_name;
	pthread_t *threads;

	/* Set the program name (pretty useless in version 1) */
	program_name = (char *) malloc(strlen(argv[0]) + 1);
	strcpy(program_name, argv[0]);

	/* Initialize our data */
	initialize_webstats();
	threads = (pthread_t *) malloc(sizeof(pthread_t) * numThreads);

	/* And we're off with our threads */
	for(i = 0; i < numThreads; i++)
		pthread_create(&threads[i], NULL, process_file, (void*) argv[i + 1]);

	/* Wait on our threads to finish up */
	for(i = 0; i < numThreads; i++)
		pthread_join(threads[i], NULL);

	/* Throw out the data */
	print_webstats();

	/* Keep it clean */
	free(program_name);
	free(threads);
	exit(0);
}

/* vim: set ts=4: */

File: version1/Webstats.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.StringTokenizer;

/**
 * Original Author: Jeff Cope
 * Maintaining Author: Amit Jain
 * Modified by: Stephen Porter
 * Class: CS 453
 */
public class Webstats {
	private static final int LBYTES = 2;
	private static final int TBYTES = 4;
	private static final int LGETS = 8;
	private static final int TGETS = 16;
	private static final int FGETS = 32;
	private static final int LFGETS = 64;
	
	private static final int ADDRESS_FIELD = 0;
	private static final int DATE_FIELD = 3;
	private static final int HTTP_STATUS_CODE_FIELD = 8;
	private static final int BYTES_DOWNLOADED_FIELD = 9;
	
	private static final String DELIMETER = " []\"";

	private double local_bytes;
	private double total_bytes;
	private double local_gets;
	private double total_gets;
	private double failed_gets;
	private double local_failed_gets;

	/**
	 * Webstat(): Initializes the Webstat structure.
	 * @param new_program_name This programs name.
	 * @return none
	 */
	public Webstats(String new_program_name) {
		local_bytes = 0;
		total_bytes = 0;
		local_gets = 0;
		total_gets = 0;
		failed_gets = 0;
		local_failed_gets = 0;
	}

	/**
	 * parse_line(): Parse one line into String tokens separated by the given
	 * delimiters.
	 * @param line A character array containing the current server log entry line
	 * @param delim A character array containing the delimiters to be used to separate fields in the line
	 * @return The number of tokens found in the line.
	 */
	private String parse_line(String line, String delim, double[] cache) {
		StringTokenizer strtok = new StringTokenizer(line, delim);
		int cnt, bytes_downloaded;
		String next, address, httpStatusCode, date;
		
		cnt = 0;
		address = httpStatusCode = date = "";

		cache[TGETS]++;

		while (strtok.hasMoreTokens()) {
			next = strtok.nextToken();
			bytes_downloaded = 0;

			if (cnt == ADDRESS_FIELD) 
				address = next;

			else if (cnt == DATE_FIELD) 
				date = next;

			else if (cnt == HTTP_STATUS_CODE_FIELD) {
				httpStatusCode = next;

				if (httpStatusCode.equals("404"))
					cache[FGETS]++;
			}

			else if (cnt == BYTES_DOWNLOADED_FIELD) {
				try {
					bytes_downloaded = Integer.parseInt(next);
				}
				catch (NumberFormatException e) {
					/* skip lines without downloaded bytes field */
				}

				cache[TBYTES] += bytes_downloaded;

				if ((address.indexOf("boisestate.edu") != -1) || (address.indexOf("132.178") != -1)) {
					cache[LGETS]++;
					cache[LBYTES] += bytes_downloaded;

					if (httpStatusCode.equals("404"))
						cache[LFGETS]++;
				}
				
				return date;
			}
			
			cnt++;
		}

		return date;
	}

	/**
	 * update_Webstat(): Updates the Webstat structure based on the input fields
	 * of current line.
	 * @param num The number of fields in the current webserver log entry
	 * @param field An array of num Strings representing the log entry
	 * @return none
	 */
	private synchronized void update_webstats(double[] cache) {
		local_bytes += cache[LBYTES];
		total_bytes += cache[TBYTES];
		local_gets += cache[LGETS];
		total_gets += cache[TGETS];
		failed_gets += cache[FGETS];
		local_failed_gets += cache[LFGETS];
	}

	/**
	 * print_webstats(): Print out webstats on standard output.
	 * @return none
	 */
	void print_webstats() {
		System.out.printf("%10s %15s   %15s  %15s\n", "TYPE", "gets", "failed gets", "MB transferred");
		System.out.printf("%10s  %15.0f  %15.0f  %15.0f\n", "local", local_gets, local_failed_gets, (double) local_bytes / (1024 * 1024));
		System.out.printf("%10s  %15.0f  %15.0f  %15.0f\n", "total", total_gets, failed_gets, (double) total_bytes / (1024 * 1024));
	}

	/**
	 * process_file(): The main function that processes one log file
	 * @param ptr Pointer to log file name.
	 * @return none output: none
	 */
	public void process_file(Object ptr) {
		double[] cache = new double[LFGETS + 1];
		String filename = (String) ptr;
		String end_date = "";

		try {
			BufferedReader fin = new BufferedReader(new FileReader(new File(filename)));
			String linebuffer = fin.readLine();
			
			if (linebuffer != null) {
				System.out.println("Starting date: " + parse_line(linebuffer, DELIMETER, cache));

				while ((linebuffer = fin.readLine()) != null)
					end_date = parse_line(linebuffer, DELIMETER, cache);
				
				System.out.println("Ending date: " + end_date);
			}

			update_webstats(cache);
			fin.close();
		}
		catch (FileNotFoundException e) {
			System.err.println("Cannot read from file " + filename);
			System.exit(1);
		}
		catch (IOException e) {
			System.err.println("Cannot open file " + filename);
			System.exit(1);
		}
	}

	/**
	 * Main insertion (Such bad Java coding)
	 */
	public static void main(String argv[]) throws InterruptedException {
		if (argv.length < 1) {
			System.err.println("Usage: java Webstats <access_log_file> {<access_log_file>}");
			System.exit(1);
		}

		Thread[] threads = new Thread[argv.length];
		Webstats ws = new Webstats("Webstats.java");

		for (int i = 0; i < argv.length; i++)
			threads[i] = createThread(ws, argv[i]);

		for (int i = 0; i < argv.length; i++)
			threads[i].join();

		ws.print_webstats();
		System.exit(0);
	}

	/**
	 * I hate doing threads this way in Java, but I figured I would keep
	 * it as close to the C solution as possible in case you didn't want
	 * any extra classes.
	 * @param ws The webstats object
	 * @param ptr The file path to kick off the process file command
	 * @return The thread
	 */
	private static Thread createThread(Webstats ws, Object ptr) {
		Thread t = new Thread(new Runnable() {
			public void run() {
				ws.process_file(ptr);
			}
		});

		t.start();
		return t;
	}
}

/* vim: set ts=4: */
